Issue464 cube metadata

VictorVerhaert opened the pull request:
from issue #464

VictorVerhaert commented:
https://github.com/Open-EO/openeo-python-client/blob/498c4c7676f46d4ed6e858178cc08b1189a7699f/openeo/metadata.py#L231-L242
Currently functionality does not change due to this function.
As this function checks the type and returns an object of the same class, functions like `reduce_dimension` (who use this function to update a metadata object) will still return a CollectionMetadata object if that was the original object.

Changing this to return a CubeMetadata changes functionality and fails some existing tests.

An alternative is to provide an optional argument to set the expected returned class, however this feels more like a quick fix and might be confusing in the future.

soxofaan reviewed (commented):
some initial comments

soxofaan commented:
> Currently functionality does not change due to this function.

I think it's fine to keep the original type of the metadata, that will indeed avoid breakage at other places.

Priority at the moment is having a base metadata class that is not implicitly tied to collection metadata (e.g. for https://github.com/Open-EO/openeo-python-client/issues/425 https://github.com/Open-EO/openeo-python-client/issues/516 https://github.com/Open-EO/openeo-python-client/issues/527 etc)

VictorVerhaert commented on openeo/metadata.py:
```diff
@@ -228,121 +217,37 @@ def __init__(self, metadata: dict, dimensions: List[Dimension] = None):
     def _clone_and_update(
         self, metadata: dict = None, dimensions: List[Dimension] = None, **kwargs
-    ) -> CollectionMetadata:
+    ) -> CubeMetadata:  # python >= 3.11: -> Self to be more correct for subclasses
```
removed the comment as this won't be relevant anytime soon

VictorVerhaert commented:
_orig_metadata has been removed from CubeMetadata.
This causes CollectionMetadata to override the _clone_and_update method in order to preserve the orig_metadata (CollectionMetadata cannot be initiated without metadata, as it should be imo). 

Unless we change _clone_and_update to always return a CubeMetadata this is the only clean solution I think.

CubeMetadata tests still need to be written.

soxofaan commented on openeo/metadata.py:
```diff
@@ -199,149 +199,45 @@ def rename_labels(self, target, source) -> Dimension:
+        if dimensions is not None:
+            self.set_dimensions(dimensions=dimensions)
+
+    def set_dimensions(self, dimensions: List[Dimension]):
```
I don't think `set_dimensions` should be a public method. 
Ideally, the metadata classes should be immutable from the outside, so they should not have setters

soxofaan replied:
I would also just keep this logic under `__init__`, which will make it easier to understand

VictorVerhaert replied:
reason I took this out of __init__ is so that the init of CollectionMetadata could reuse this code after parsing dimensions from metadata. 
I added some todo's about type checking e.g. BandsDimension, that would need to be resolved on two places otherwise.

soxofaan commented on openeo/metadata.py:
```diff
@@ -199,149 +199,45 @@ def rename_labels(self, target, source) -> Dimension:
+            self.set_dimensions(dimensions=dimensions)
+
+    def set_dimensions(self, dimensions: List[Dimension]):
+        if dimensions == None:
```
FYI: python tip: equality to `None` should be done with `is None` instead of `== None`

soxofaan reviewed (commented):
some quick notes

VictorVerhaert commented:
TODO:
- [ ] Write test cases specific for CubeMetadata

soxofaan commented on openeo/metadata.py:
```diff
@@ -84,6 +84,10 @@ def __init__(self, name: str, extent: Union[Tuple[str, str], List[str]]):
         return TemporalDimension(name=name, extent=self.extent)
 
+    def rename_labels(self, target, source) -> Dimension:
+        # TODO should we check if the extend has changed with the new labels?
```
```suggestion
        # TODO should we check if the extent has changed with the new labels?
```

soxofaan reviewed (commented):
some more notes

about test_metadata.py: be careful about adding formatting changes everywhere, that makes it hard to to find the actual things that were changed

can you also add a changelog entry?

VictorVerhaert commented:
Because I changed the names of the metadata tests using CollectionMetadata to test_collectionmetadata, pre-commit refactored the whole tests causing a large diff.
I'll be more careful with this in the future.

I added an entry in the changelog and processed the last review.

soxofaan commented on CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 - Changed default `chunk_size` of various `download` functions from None to 10MB. This improves the handling of large downloads and reduces memory usage. ([#528](https://github.com/Open-EO/openeo-python-client/issues/528))
 - `Connection.execute()` and `DataCube.execute()` now have a `auto_decode` argument. If set to True (default) the response will be decoded as a JSON and throw an exception if this fails, if set to False the raw `requests.Response` object will be returned. ([#499](https://github.com/Open-EO/openeo-python-client/issues/499))
+- Created superclass `CubeMetadata` for `CollectionMetadata` containing methods to handle basic metadata like dimensions and their labels, without requiring STAC information. ([#464](https://github.com/Open-EO/openeo-python-client/issues/464))
```
```suggestion
- Introduced superclass `CubeMetadata` for `CollectionMetadata` for essential metadata handling (just dimensions for now) without collection-specific STAC metadata parsing. ([#464](https://github.com/Open-EO/openeo-python-client/issues/464))
```
